src/gstring.rs

```diff
@@ -0,0 +1,440 @@
+
+    fn deref(&self) -> &str {
+        self.as_str()
+    }
+}
+
+impl From<GString> for String {
+    #[inline]
+    fn from(s: GString) -> Self {
+        std::mem::forget(s.ptr);
+        String::from(s.as_str())
+    }
+}
+
+impl From<GString> for Box<str> {
+    #[inline]
+    fn from(s: GString) -> Self {
+        std::mem::forget(s.ptr);
+        Box::from(s.as_str())
+    }
+}
+
+impl From<String> for GString {
+    #[inline]
+    fn from(s: String) -> Self {
+        GString::from(s.as_bytes().to_vec())
+    }
+}
+
+impl From<Box<str>> for GString {
+    #[inline]
+    fn from(s: Box<str>) -> Self {
+        s.as_bytes().to_vec().into()
+    }
+}
+
+impl From<&'static str> for GString {
+    #[inline]
+    fn from(s: &'static str) -> Self {
+        s.as_bytes().to_vec().into()
+    }
+}
+
+impl From<Vec<u8>> for GString {
+    #[inline]
+    fn from(s: Vec<u8>) -> Self {
+        let cstr = unsafe { CString::from_vec_unchecked(s) };
+        Self::from(cstr)
+    }
+}
+
+impl From<CString> for GString {
+    #[inline]
+    fn from(s: CString) -> Self {
+        Self { s: Some(s), ptr: ptr::null(), length: 0 }
+    }
+}
+
+impl From<&CStr> for GString {
+    #[inline]
+    fn from(c: &CStr) -> Self {
+        let cstring = CString::from(c);
+        Self::from(cstring)
+    }
+}
+
+impl FromGlibPtrFull<*const c_char> for GString {
+    #[inline]
+    unsafe fn from_glib_full(ptr: *const c_char) -> Self {
+        let dst = std::ptr::NonNull::dangling();
+        std::ptr::write(dst.as_ptr(), std::ptr::read(ptr));
+        GString::new(dst.as_ptr() as *const c_char)
+    }
+}
+
+impl FromGlibPtrFull<*mut u8> for GString {
+    #[inline]
+    unsafe fn from_glib_full(ptr: *mut u8) -> Self {
+        let dst = std::ptr::NonNull::new_unchecked(ptr);
+        std::ptr::write(dst.as_ptr(), std::ptr::read(ptr));
+        GString::new(dst.as_ptr() as *const c_char)
+    }
+}
+
+impl FromGlibPtrFull<*mut i8> for GString {
+    #[inline]
+    unsafe fn from_glib_full(ptr: *mut i8) -> Self {
+        let dst = std::ptr::NonNull::new_unchecked(ptr);
+        std::ptr::write(dst.as_ptr(), std::ptr::read(ptr));
+        GString::new(dst.as_ptr() as *const c_char)
+    }
+}
+
+impl FromGlibPtrNone<*const c_char> for GString {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *const c_char) -> Self {
+        let cstr = CStr::from_ptr(ptr);
+        cstr.into()
+    }
+}
+
+impl FromGlibPtrNone<*mut u8> for GString {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *mut u8) -> Self {
+        let cstr = CStr::from_ptr(ptr as *mut i8);
+        cstr.to_owned().into()
+    }
+}
+
+impl FromGlibPtrNone<*mut i8> for GString {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *mut i8) -> Self {
+        let cstr = CStr::from_ptr(ptr);
+        cstr.to_owned().into()
+    }
+}
+
+impl FromGlibPtrBorrow<*const c_char> for GString {
+    #[inline]
+    unsafe fn from_glib_borrow(ptr: *const c_char) -> Self {
+        GString::new(ptr)
```

Comment:
You need to mark this somehow as borrowed in the `GString` so that the `Drop` impl is not freeing the pointer. That's why an enum would be best for the representation of it

Reply:
Yes I started moving into that direction...